### PR TITLE
Startup hooks as browser slots

### LINKED ISSUE #2665: Move *after-startup-hook* and *after-startup-hook* to browser slots
If we take a look at `global.lisp`, we  see that beyond the REPL / dev options, there are only 4 variables that impact the global state:

- `*options*`
- `*global*`
- `*after-startup-hook*`
- `*after-init-hook*`

The less the better.  So what about moving  `*after-startup-hook*` and `*after-startup-hook*` to browser slots?  Seems pretty natural since they can only be used in a context where `*browser*` exists.  

I don't know why they are not there already, maybe an artifact from a very old version of Nyxt.

Thoughts?

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This moves the clunky `*after-(init|startup)-hook*` into `browser` slots.

Fixes #2665.

# Discussion

I don't like that in our testing code we `nyxt:start` and then set the handlers. While it works for now, it's terribly temporarity-dependent and is unlikely to work forever.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [X] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [X] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [X] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [X] I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - [X] (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [X] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by Ambrevar** [comment]
> #
> 
> I don't like that in our testing code we `nyxt:start` and then set the handlers. While it works for now, it's terribly temporarity-dependent and is unlikely to work forever.

Yes, this is clunky.  So instead I propose to make the offending tests a bit more usable: instead of calling `start`, let's instantiate browser ourselves, set the hook slots and run `finalize` to start the browser.

Thoughts?

**Comment by aartaka** [comment]
> > I don't like that in our testing code we `nyxt:start` and then set the handlers. While it works for now, it's terribly temporarity-dependent and is unlikely to work forever.
> 
> Yes, this is clunky. So instead I propose to make the offending tests a bit more usable: instead of calling `start`, let's instantiate browser ourselves, set the hook slots and run `finalize` to start the browser.
> 
> Thoughts?

Yes, that is a good idea. But let's do that after we revive `*global-profile*` and make `headless-p` a browser slot.

**Comment by aadcg** [comment]
I agree with everything here. It needs to be rebased and it should be ready to be merged.

**Review by aadcg — APPROVED** [review]

**Comment by aartaka** [comment]
@aadcg, rebased :)

**Comment by aadcg** [comment]
When I mentioned rebasing, it wasn't a mere rebase over master. Notice that `with-headless` doesn't exist anymore for instance.

**Comment by aadcg** [comment]
I can do it myself also, let me know.

**Comment by aartaka** [comment]
Should be fine now.

**Comment by aartaka** [comment]
Waiting for CI to (not) fail, but looks like I managed to do it myself :P

**Comment by Ambrevar** [comment]
OK with me.

To test the startup procedure, we could  (in another PR), split the tests in 2:
- To test `entry-point` and `start`:
  - either add an option to pass a handler after init;
  - or have `start` reuse an existing `*browser*`;
  - or (best option in my opinion), have a client listen to the socket for readiness, then send instruction to quit.
- To test all other browser initialization features, bypass `start` and initialize `browser` ourselves.

File: tests/renderer-offline/execute-command-eval.lisp [hunk 1]
```diff
@@ -4,26 +4,30 @@
...
-    (nyxt:start :no-config t :no-auto-config t
-                :headless t
-                :socket "/tmp/nyxt-test.socket"
-                :profile "test")
-    (let ((result (calispel:? channel)))
-      (assert-eq '1+ (first result))
-      (assert-eq 3 (second result))))
-  (sleep 1)
-  (nyxt:quit))
+  (let ((old-headless-p nyxt::*headless-p*))
```
Comment (aadcg) [inline tests/renderer-offline/execute-command-eval.lisp:7]: I don't understand these changes. Let-binding these global variables was removed in the past.

Comment (aartaka) [reply tests/renderer-offline/execute-command-eval.lisp:7]: Damn, that's where I've gotten lost :D

**Comment by aadcg** [comment]
I'll clean this up.

**Comment by aadcg** [comment]
@Ambrevar detailed a plan to test out startup procedure (I'll open an issue about that).

There is a blocker to merge this PR. Running `(asdf:test-system "nyxt/gi-gtk")` fails. See 9833c5949.

File: tests/renderer-offline/execute-command-eval.lisp (hunk 1 above)
Comment (aadcg) [reply tests/renderer-offline/execute-command-eval.lisp]: It's absolutely understandable. Some of these git gymnastics needs to be done manually.

**Comment by aartaka** [comment]
@aadcg, yet another rebase mistake of mine (^_^;) I've moved `nyxt:start` earlier in the test and fixes `test-set-url`, should work now.

**Comment by aadcg** [comment]
Or perhaps my mistake! Thanks @aartaka.

**Comment by aartaka** [comment]
> Or perhaps my mistake! Thanks @aartaka.

No, it's mine :P The `test-set-url` was altered in parallel to this branch, so it was not refactored here.

**Comment by aartaka** [comment]
Thanks for checking thouroughly, @aadcg!

**Pull request merged**